rust/worker/src/execution/orchestration/get.rs

```diff
@@ -0,0 +1,327 @@
+    ) -> Self {
+        Self {
+            dispatcher,
+            queue,
+            fetch_log,
+            fetch_segment,
+            fetch_log_output: None,
+            fetch_segment_output: None,
+            filter,
+            limit,
+            projection,
+            result_channel: None,
+        }
+    }
+
+    pub async fn run(mut self, system: System) -> GetResult {
+        let (tx, rx) = oneshot::channel();
+        self.result_channel = Some(tx);
+        let mut handle = system.start_component(self);
+        let result = rx.await;
+        handle.stop();
+        result?
+    }
+
+    // Cleanup the task result and produce the output if any
+    // Terminate the orchestrator if there is any error
+    fn cleanup_response<O, E>(
+        &mut self,
+        ctx: &ComponentContext<Self>,
+        message: TaskResult<O, E>,
+    ) -> Option<O>
+    where
+        E: Into<GetError>,
+    {
+        match message.into_inner() {
+            Ok(output) => Some(output),
+            Err(err) => {
+                self.terminate_with_error(ctx, err.into());
+                None
+            }
+        }
+    }
+
+    fn terminate_with_error(&mut self, ctx: &ComponentContext<Self>, err: GetError) {
+        tracing::error!("Error running orchestrator: {}", err);
+        terminate_with_error(self.result_channel.take(), err, ctx);
+    }
+
+    // Sends the task to dispatcher and returns whether the action is successful
+    // Terminate the orchestrator if there is any error
+    async fn send_task(&mut self, ctx: &ComponentContext<Self>, task: TaskMessage) -> bool {
+        if let Err(err) = self.dispatcher.send(task, Some(Span::current())).await {
+            self.terminate_with_error(ctx, err.into());
+            false
+        } else {
+            true
+        }
+    }
+
+    async fn try_start_filter_operator(&mut self, ctx: &ComponentContext<Self>) {
+        if let (Some(logs), Some(segments)) = (
+            self.fetch_log_output.as_ref(),
+            self.fetch_segment_output.as_ref(),
+        ) {
+            let task = wrap(
+                Box::new(self.filter.clone()),
+                FilterInput {
+                    logs: logs.clone(),
+                    segments: segments.clone(),
+                },
+                ctx.receiver(),
+            );
+            self.send_task(ctx, task).await;
+        }
+    }
+}
+
+#[async_trait]
+impl Component for GetOrchestrator {
+    fn get_name() -> &'static str {
+        "Get Orchestrator"
+    }
+
+    fn queue_size(&self) -> usize {
+        self.queue
+    }
+
+    async fn on_start(&mut self, ctx: &ComponentContext<Self>) {
+        let log_task = wrap(Box::new(self.fetch_log.clone()), (), ctx.receiver());
+        let segment_task = wrap(Box::new(self.fetch_segment.clone()), (), ctx.receiver());
+        let log_task_success = self.send_task(ctx, log_task).await;
+        if log_task_success {
+            self.send_task(ctx, segment_task).await;
```

Comment:
nit - comment that we are bailing for error

Reply:
I've refactored a bit here. Hopefully that's clearer now.